datalad_catalog/workflows.py

```diff
@@ -0,0 +1,248 @@
+    # Create catalog
+    cat = catalog
+    if not cat.is_created():
+        cat.create()
+    # Call per-dataset workflow
+    def _dataset_workflow_inner(ds, refds, **kwargs):
+        """Internal function to allow passing"""
+        return dataset_workflow(ds, catalog=cat, **kwargs)
+
+    try:
+        for res in ds.foreach_dataset(
+            _dataset_workflow_inner,
+            recursive=True,
+            recursion_limit=1,
+            state="any",
+            return_type="generator",
+            on_failure="continue",
+        ):
+            # unwind result generator
+            for partial_result in res.get("result", []):
+                yield partial_result
+    except IncompleteResultsError as e:
+        print(
+            f"Could not run workflow for all datasets. Inspect errors:\n\n{e}"
+        )
+
+    # Set super dataset of catalog
+    cat.main_id = ds.id
+    # - sync possible adjusted branch
+    ds.repo.localsync()
+    # - account for possibility of being on adjusted branch:
+    cat.main_version = ds.repo.get_hexsha(ds.repo.get_corresponding_branch())
+    cat.set_main_dataset()
+
+
+def update_workflow(superds_path, subds_path, catalog: WebCatalog):
+    """Run an update workflow on a specific subdataset and its parent
+
+    The workflow includes running several steps on the super- and subdataset:
+    - dataset- and file-level metadata extraction
+    - extracted metadata translation
+    - adding translated metadata to the catalog
+
+    It then resets the catalog's superdataset to the latest id and version of
+    the parent dataset.
+
+    This workflow assumes:
+    - The subdataset has already been added as a submodule to the parent dataset
+    - The parent dataset already contains the subdataset commit
+    """
+    # Install super and subdatasets
+    super_ds = Dataset(superds_path)
+    sub_ds = Dataset(subds_path)
+    # Dataset workflow for super+subdataset
+    yield dataset_workflow(super_ds, catalog)
+    yield dataset_workflow(sub_ds, catalog)
+    # Set super dataset of catalog
+    catalog.main_id = super_ds.id
+    # - sync possible adjusted branch
+    super_ds.repo.localsync()
+    # - account for possibility of being on adjusted branch:
+    catalog.main_version = super_ds.repo.get_hexsha(
+        super_ds.repo.get_corresponding_branch()
+    )
+    catalog.set_main_dataset()
+
+
+def dataset_workflow(ds: Dataset, catalog, **kwargs):
+    """Run a dataset-specific catalog generation workflow.
+
+    This includes:
+      - dataset- and file-level metadata extraction
+      - extracted metadata translation
+      - adding translated metadata to a catalog"""
+    # 1. Run dataset-level extraction
+    extracted_file = Path(ds.path) / "extracted_meta.json"
+    for name in extractor_names_dataset:
+        if check_required_files(ds, name):
+            metadata_record = extract_dataset_level(ds, name)
+            write_jsonline_to_file(extracted_file, metadata_record)
```

Comment:
Currently, one of the possible extractor names is bids_dataset, but this extractor is more of an "optional dependency" for the catalog (it's not part of datalad-catalog, datalad-metalad or other formally declared dependencies). I think a warning would be appropriate if the extractor is missing (and I wouldn't introduce a strict dependency on -neuroimaging).

Reply:
Agreed